apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -72,6 +81,61 @@ store_message(Msg) ->
 find_subscribers(_Msg) ->
     [node()].
 
+open_session(ClientID) ->
+    ?WHEN_ENABLED(emqx_ds:session_open(ClientID)).
+
+-spec add_subscription(emqx_types:topic(), emqx_ds:session_id()) ->
+    {ok, emqx_ds:iterator_id(), IsNew :: boolean()} | {skipped, disabled}.
+add_subscription(TopicFilterBin, DSSessionID) ->
+    ?WHEN_ENABLED(
+        begin
+            TopicFilter = emqx_topic:words(TopicFilterBin),
+            {ok, IteratorID, StartMS, IsNew} = emqx_ds:session_add_iterator(
```

Comment:
I think we should get `StartMS` on that level, and free `emqx_ds` level from dealing with current time. Consider this planned feature where clients can ask to start the replay at some point in the past.

Reply:
Missed that idempotency considerations will make `emqx_ds` hairier. Disregard the comment.